### ISSUE TITLE
Tracking Issue for `mem::conjure_zst` (`feature(mem_conjure_zst)`)

### ISSUE THREAD

**Issue by scottmcm** [rust-lang/rust#95383]
Feature gate: `#![feature(mem_conjure_zst)]`

This is a tracking issue for the `mem::conjure_zst::<T>()` function.

There are a variety of possible ways one can do this without this function, but it provides two advantages:

1) It makes it clearer to the reader what's happening and keeps the code author from needing to decide between `zeroed()` or `uninitialized()` or `MaybeUninit::new_uninit().assume_init()` in situations like
https://github.com/rust-lang/rust/blob/ab0c2e18dceb7140626a158affb983ae81039bd0/library/alloc/src/vec/into_iter.rs#L149-L150

2) It provides a convenient place to talk about why it's not just *safe* to do this in all cases.

### Public API

```rust
// core::mem

pub const unsafe fn conjure_zst<T>() -> T;
```

### Steps / History

- [x] ACP: https://github.com/rust-lang/libs-team/issues/292#issuecomment-2532315581
- [x] Library Implementation:
    - #95385
    - #146479
- [ ] Lint Implementation: *todo*
- [ ] Final comment period (FCP)
- [ ] Stabilization PR

### Unresolved Questions

- Should we have a lint suggesting that other ways of creating a ZST, particularly things like `*ptr::null::<()>()` (see https://github.com/rust-lang/rust/pull/148122#issuecomment-3462289908), should be changed to call this instead (once it's stablized)?

**Comment by clarfonthey** [comment]
Posting this here since it doesn't have to be part of the initial PR, but I think there should be a lint for creating uninitialised ZSTs using other methods, since this should be the canonical way to do so.

Another potential thing: if we make it so that calling `conjure_zst` is a compiler error for non-ZSTs, then what exactly is the `unsafe` signifying? We can't actually cause any undefined behaviour with it, and although it's currently not possible to do this in safe code, I don't think it's necessarily bad to make it safe, although I do think that it shouldn't be used outside unsafe code. Definitely could break invariants with it.

**Comment by scottmcm** [comment]
> Definitely could break invariants with it.

That's *why* it has to be unsafe -- see the docs in the PR.

I can write code whose soundness depends on the fact that the only way to get an instance of a particular ZST is by asking me.  For example, maybe when you ask for the ZST I initialize some static state.  Then even if the ZST is `Copy`, I can require that you pass me an instance of it and I know that said state is already initialized, without needing to check.

But if you can just conjure that ZST in safe code, that would become unsound.

> if we make it so that calling conjure_zst is a compiler error for non-ZSTs

Note that it being an *error* to even mention for a non-ZST would make it not work in the kinds of places that inspired me to make it.  If you look at the PR, the uses are guarded by *runtime* zero-sizedness checks -- the compiler doesn't know it's a ZST.  And it's monomorphized for non-ZSTs, those just happen to be in dead code.

**Comment by clarfonthey** [comment]
So, I went through a bit of a journey looking into this, and I'll share what's relevant for this discussion.

First, the assertion intrinsics. Right now, we have the following intrinsics which add unconditional panics to cases of methods being used incorrectly. Specifically:

* `intrinsics::assert_zero_valid` => `mem::zeroed`
* `intrinsics::assert_mem_uninitialized_valid` => `mem::uninitialized`
* `intrinsics::assert_inhabited` => `mem::MaybeUninit::assume_init`

I went through adding this in #147287 thinking that this was the mechanism used for the lint, without actually checking. It turns out these exist *just* for the unconditional panics, and for that, we can use `assert_inhabited` and a UB assertion for `size_of::<T>() == 0` to avoid bloating the code.

Now, the main lint that seems relevant here is `invalid_value`, which is warn-by-default instead of deny-by-default. This is defined in `compiler/rustc_lint/src/builtin.rs` and looks for calls to:

* `mem::zeroed`
* `mem::uninitialized`
* `mem::transmute` (if value is zero, this treats like `mem::zeroed`)
* `mem::MaybeUninit::assume_init`

Based upon what the libs team said, it's a bit weird that this lint is warn-by-default, and I would assume that this is solely because deny-by-default was causing issues, but I don't actually see any of that searching for `invalid_value` in the repo.

Basically, I see one of two paths forward:

1. Propose making `invalid_value` a deny-by-default lint and run crater to see what this breaks.
2. Add a new lint that specifically covers `conjure_zst`.

I personally think 1 is the better option, but it's worth pointing out both options here.

**Comment by RalfJung** [comment]
Also, new lints typically need t-lang approval.

**Comment by clarfonthey** [comment]
To be fair, I was under the impression initially that the way these hints worked was by knowing that `assert_inhabited` and the like would always panic, and then firing an `unconditional_panic` lint. From this perspective, I think that it would be fair to trigger the lint in these cases.

But I guess you're right: this is just a matter of an API violation, not actual UB in all cases.

**Comment by Apersoma** [comment]
can the starting comment to be edited to reference the new pr? (#146479)

**Comment by theemathas** [comment]
Given that quite some time has passed, I propose to stabilize this API. I think that the lint mentioned as an unresolved question can be done later at any point.

---

Stabilized API:

```rust
pub const unsafe fn conjure_zst<T>() -> T;
```

The notable thing about this implementation is that if this function is called at run time (as opposed to in consteval), then the function will panic at run time (as opposed to at compile time) if `T` is not a ZST.

If the function panics in consteval, the panic message is a fixed literal string. If the function panics at run time, the panic message includes the `type_name` of `T`. Note that [`type_name`](https://doc.rust-lang.org/nightly/std/any/fn.type_name.html) is currently [unstable in const](https://github.com/rust-lang/rust/issues/63084).

---

Implementation history:

* https://github.com/rust-lang/rust/pull/95385 (attempted to add the API without an ACP, and then the PR was closed)
* https://github.com/rust-lang/libs-team/issues/292 (The ACP. It has a lot of additional context and discussion, including the explanation of why this is needed)
* https://github.com/rust-lang/rust/pull/146479 (the actual implementation)
* https://github.com/rust-lang/rust/pull/153073 (changes the panic message)
* https://github.com/rust-lang/rust/pull/155840 (internal implementation change)

**Comment by clarfonthey** [comment]
We'd want to stabilise on a PR, if you're fine making one.
